File: Linear Control Course/Chapter19/Lesson2/Chapter19_Lesson2.cpp
#include <iostream>
#include <Eigen/Dense>
// In a robotic context, this could be part of a ROS control loop:
// #include <ros/ros.h>

struct LagCompensator {
    double Kc;
    double beta;
    double T;
    double h;      // sampling period
    double a1;
    double b0;
    double b1;
    double u_prev;
    double e_prev;

    LagCompensator(double Kc_, double beta_, double T_, double h_)
        : Kc(Kc_), beta(beta_), T(T_), h(h_),
          u_prev(0.0), e_prev(0.0)
    {
        // Backward Euler discretization of:
        // beta T du/dt + u = Kc T de/dt + Kc e
        // After algebra, one obtains:
        double denom = beta * T / h + 1.0;
        a1 = (beta * T / h) / denom;
        b0 = (Kc * (T / h + 1.0)) / denom;
        b1 = (-Kc * (T / h)) / denom;
    }

    double update(double e) {
        double u = a1 * u_prev + b0 * e + b1 * e_prev;
        u_prev = u;
        e_prev = e;
        return u;
    }
};

int main() {
    LagCompensator lag(5.0, 5.0, 1.0, 0.001); // Kc, beta, T, h
    double e = 0.1; // example tracking error
    for (int k = 0; k < 10; ++k) {
        double u = lag.update(e);
        std::cout << "k=" << k << " u=" << u << std::endl;
    }
    return 0;
}

File: Linear_Control_Course/Chapter19/Lesson2/Chapter19_Lesson2.java
public class LagCompensator {
    private final double Kc;
    private final double beta;
    private final double T;
    private final double h;
    private final double a1;
    private final double b0;
    private final double b1;
    private double uPrev = 0.0;
    private double ePrev = 0.0;

    public LagCompensator(double Kc, double beta, double T, double h) {
        this.Kc = Kc;
        this.beta = beta;
        this.T = T;
        this.h = h;
        double denom = beta * T / h + 1.0;
        this.a1 = (beta * T / h) / denom;
        this.b0 = (Kc * (T / h + 1.0)) / denom;
        this.b1 = (-Kc * (T / h)) / denom;
    }

    public double update(double error) {
        double u = a1 * uPrev + b0 * error + b1 * ePrev;
        uPrev = u;
        ePrev = error;
        return u;
    }
}

// Example usage in a robot periodic loop (e.g., teleopPeriodic):
// LagCompensator lag = new LagCompensator(5.0, 5.0, 1.0, 0.02);
// double error = referencePosition - measuredPosition;
// double control = lag.update(error);
